### DISCUSSION TITLE
Will we be able to access more of the request in RSC?

### DISCUSSION THREAD

**Discussion by jchild3rs** [vercel/next.js#47117] [App Router]
Specifically in layout components, it's just `params` defined by the file structure routing. Then we have `headers()` and `cookies()`. What about query parameters? What about users who leverage rewrites for routing? Why just route parameters?

Maybe there is a `searchParams()` meant for the server that lives along side `headers()` and `cookies()`. It's strange that we can access this in client components with `useSearchParams()`, but not the server.

**Comment by leerob** [comment]
What are you looking to accomplish / build in your application with the request data? This would be very helpful to understand.

**Reply by m5chm3lz3r** [reply]
How can we seamlessly access the locale or full URL of the request in server-side components, and then use this information to properly render translations on the initial server render for client-side components?

Currently, with packages such as next-i18next, there is a brief switch between translations during the initial rendering of the client-side component and the eventual rerender in the browser, causing a suboptimal user experience.

One example of a workaround involves using middleware to store the locale in the headers, where there is access to the request object. From there, the server-side component can access the locale indirectly. Although a similar approach could be used with the URL, it feels like a less-than-ideal solution for what should be an essential feature.

The provided example from next-i18next demonstrates the translation switch issue during rendering. You can access it here: https://github.com/i18next/next-13-app-dir-i18next-no-locale-path-example.

Additionally, is there a method to utilize a functionality similar to getServerSideProps with client-side components? It would be beneficial to have a means of executing server-only code solely on the initial (server) rendering of the client component. To provide an example, the locale could be obtained via headers (or a request object that directly contains it) and then passed to the props as static values that remain constant during the rerender on the frontend.

It is unclear whether these features I have outlined are areas where Vercel could make improvements or if they must be addressed directly with the React team. However, I believe these are not niche features but rather crucial ones for any full-scale Next.js 13 application.

**Reply by runoncedev** [reply]
I'm looking for a way to read the URL search params from a RSC. I want to store the state of the app in the URL and render the initial HTML depending on that.

**Reply by axelvaindal** [reply]
I was wondering the same thing regarding the access of the URL in the RSC.

Perhaps an utility functions in the same spirit as `cookies()` and `headers()` could be considered, something like `url()` ?

it would at least avoid settings headers in middleware (which are edge only). We can make the data readonly and I fail to see what would be the catch here
